// @2pg/oauth/lib/types/guild.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var permissionConstants = {
    1: 'CREATE_INSTANT_INVITE',
    2: 'KICK_MEMBERS',
    4: 'BAN_MEMBERS',
    8: 'ADMINISTRATOR',
    0x10: 'MANAGE_CHANNELS',
    0x20: 'MANAGE_GUILD',
    0x40: 'ADD_REACTION',
    0x80: 'VIEW_AUDIT_LOG',
    0x400: 'VIEW_CHANNEL',
    0x800: 'SEND_MESSAGES',
    0x1000: 'SEND_TTS_MESSAGES',
    0x2000: 'MANAGE_MESSAGES',
    0x4000: 'EMBED_LINKS',
    0x8000: 'ATTACH_FILES',
    0x10000: 'READ_MESSAGES_HISTORY',
    0x20000: 'MENTION_EVERYONE',
    0x40000: 'USE_EXTERNAL_EMOJIS',
    0x100000: 'CONNECT',
    0x200000: 'SPEAK',
    0x400000: 'MUTE_MEMBERS',
    0x800000: 'MANAGE_NICKNAMES',
    0x1000000: 'MANAGE_ROLES',
    0x2000000: 'MANAGE_WEBHOOKS',
    0x4000000: 'MANAGE_EMOJIS'
};
var Guild = /** @class */ (function () {
    function Guild(_a) {
        var id = _a.id, name = _a.name, icon = _a.icon, _b = _a.features, features = _b === void 0 ? [] : _b, _c = _a.owner, owner = _c === void 0 ? false : _c, _d = _a.permissions, permissions = _d === void 0 ? 0 : _d;
        this.id = id;
        this.name = name;
        this.iconHash = icon;
        this.features = features;
        this.isOwner = owner;
        this.permissions = this.parsePermissions(permissions);
    }
    Object.defineProperty(Guild.prototype, "createdTimestamp", {
        /** The timestamp of creation of the user's account. */
        get: function () {
            return parseInt((BigInt(this.id) >> BigInt(22)).toString()) + 1420070400000;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Guild.prototype, "createdAt", {
        /** The time of creation of the user's account. */
        get: function () {
            return new Date(this.createdTimestamp);
        },
        enumerable: false,
        configurable: true
    });
    Guild.prototype.parsePermissions = function (perms) {
        var p = [];
        for (var c in permissionConstants) {
            var permNum = parseInt(c);
            if ((permNum & perms) === permNum)
                p.push(permissionConstants[permNum]);
        }
        return p;
    };
    /**
     * Returns a url to the guild icon.
     * @param size The size of the icon in pixels. (Defaults to 512)
     */
    Guild.prototype.iconUrl = function (size) {
        if (size === void 0) { size = 512; }
        return this.iconHash
            ? "https://cdn.discordapp.com/icons/" + this.id + "/" + this.iconHash + "." + (this.iconHash.startsWith('a_') ? 'gif' : 'png') + "?size=" + size
            : 'https://i.imgur.com/LvroChs.png';
    };
    return Guild;
}());
exports.default = Guild;
