// @2pg/oauth/lib/types/user.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/** A discord user who has authorized your app to have access to their data. */
var User = /** @class */ (function () {
    function User(_a) {
        var username = _a.username, locale = _a.locale, mfa_enabled = _a.mfa_enabled, _b = _a.flags, flags = _b === void 0 ? 0 : _b, _c = _a.avatar, avatar = _c === void 0 ? null : _c, discriminator = _a.discriminator, id = _a.id, _d = _a.email, email = _d === void 0 ? undefined : _d, _e = _a.verified, verified = _e === void 0 ? undefined : _e, _f = _a.premium_type, premium_type = _f === void 0 ? 0 : _f, _g = _a.bot, bot = _g === void 0 ? false : _g;
        this.username = username;
        this.locale = locale;
        this.isMFAEnabled = mfa_enabled;
        this.discriminator = parseInt(discriminator).toString().padStart(4, '0');
        this.id = id;
        this.emailId = email;
        this.emailVerified = verified;
        this.avatarHash = avatar;
        this.userFlags = [];
        this.premiumType = premium_type === 0 ? 'None' : premium_type === 1 ? 'Nitro Classic' : 'Nitro';
        this.bot = bot;
        this.displayAvatarURL = this.avatarURL({ dynamic: true, size: 256 });
        this.tag = this.username + "#" + this.discriminator;
        this.buildFlags(flags);
    }
    User.prototype.buildFlags = function (flags) {
        if ((flags & 1) === 1)
            this.userFlags.push('Discord Employee');
        if ((flags & 2) === 2)
            this.userFlags.push('Discord Partner');
        if ((flags & 4) === 4)
            this.userFlags.push('HypeSquad Events');
        if ((flags & 8) === 8)
            this.userFlags.push('Bug Hunter Level 1');
        if ((flags & 64) === 64)
            this.userFlags.push('HypeSquad House of Bravery');
        else if ((flags & 128) === 128)
            this.userFlags.push('HypeSquad House of Brilliance');
        else if ((flags & 256) === 256)
            this.userFlags.push('HypeSquad House of Balance');
        if ((flags & 512) === 512)
            this.userFlags.push('Early Supporter');
        if ((flags & 1024) === 1024)
            this.userFlags.push('Team User');
        if ((flags & 4096) === 4096)
            this.userFlags.push('System');
        if ((flags & 16384) === 16384)
            this.userFlags.push('Bug Hunter Level 2');
        if ((flags & 131072) === 131072)
            this.userFlags.push('Verified Bot Developer');
    };
    Object.defineProperty(User.prototype, "createdTimestamp", {
        /** The timestamp of the creation of the user's account. */
        get: function () {
            return parseInt((BigInt(this.id) >> BigInt(22)).toString()) + 1420070400000;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(User.prototype, "createdAt", {
        /** The time of creation of the user's account. */
        get: function () {
            return new Date(this.createdTimestamp);
        },
        enumerable: false,
        configurable: true
    });
    /** Get the URL of a user's avatar, with options. */
    User.prototype.avatarURL = function (options) {
        var _a;
        if (options === void 0) { options = { size: 512 }; }
        var extension = (((_a = this.avatarHash) === null || _a === void 0 ? void 0 : _a.startsWith('a_')) && options.dynamic) ? 'gif' : 'png';
        return "https://cdn.discordapp.com/" + (this.avatarHash ? '' : 'embed/') + "avatars/" + (this.avatarHash ? this.id + "/" + this.avatarHash : parseInt(this.discriminator) % 5) + "." + ((this.avatarHash) ? extension : 'png') + "?size=" + options.size;
    };
    return User;
}());
exports.default = User;
